pkg/controllers/management/clusterprovisioner/helm_cleaner.go

```diff
@@ -0,0 +1,297 @@
+func removeResourceWithNoSuccessors(count int, manifest string, deprecatedAPI string) string {
+	for repl := 0; repl < count; repl++ {
+		// find the position where the API header is
+		apiIndex := strings.Index(manifest, deprecatedAPI)
+
+		// find the next separator index
+		separatorIndex := strings.Index(manifest[apiIndex:], "---\n")
+
+		// find the previous separator index
+		previousSeparatorIndex := strings.LastIndex(manifest[:apiIndex], "---\n")
+
+		/*
+		 * if no previous separator index was found, it means the resource is at the beginning and not
+		 * prefixed by ---
+		 */
+		if previousSeparatorIndex == -1 {
+			previousSeparatorIndex = 0
+		}
+
+		if separatorIndex == -1 { // this means we reached the end of input
+			manifest = manifest[:previousSeparatorIndex]
+		} else {
+			manifest = manifest[:previousSeparatorIndex] + manifest[separatorIndex+apiIndex:]
+		}
+	}
+
+	manifest = strings.Trim(manifest, "\n")
+	return manifest
+}
+
+// updateRelease updates a release in the cluster with an equivalent with the superseded APIs replaced or removed as
+// needed.
+// Logic extracted from https://github.com/helm/helm-mapkubeapis/blob/main/pkg/v3/release.go#L71-L94
+func updateRelease(originalRelease *release.Release, modifiedManifest string, config *action.Configuration) error {
+	originalRelease.SetStatus(release.StatusSuperseded, "")
+	if err := config.Releases.Update(originalRelease); err != nil {
+		return errors.Wrapf(err, "[updateRelease] failed to update original release %v in namespace %v", originalRelease.Name, originalRelease.Namespace)
+	}
+
+	newRelease := copyReleaseData(originalRelease, modifiedManifest, config.Now())
+
+	logrus.Infof("[updateRelease] add release version %v for release %v with updated supported APIs in namespace %v", originalRelease.Version, originalRelease.Name, originalRelease.Namespace)
+
+	if err := config.Releases.Create(newRelease); err != nil {
+		originalRelease.SetStatus(release.StatusDeployed, "")
+		updateErr := config.Releases.Update(newRelease)
+		if updateErr != nil {
+			return errors.Wrapf(err, "[updateRelease] failed to create new release version %v for release %v in namespace %v and failed to rollback to previous version", newRelease.Version, newRelease.Name, newRelease.Namespace)
+		}
+
+		return errors.Wrapf(err, "[updateRelease] failed to create new release version %v for release %v in namespace %v", newRelease.Version, newRelease.Name, newRelease.Namespace)
+	}
+
+	logrus.Infof("[updateRelease] successfully created new version for release %v in namespace %v", newRelease.Name, newRelease.Namespace)
+
+	return nil
+}
+
+// copyReleaseData copies the original release data into a new object in order to retain previous release information
+// for rollback, if needed.
+func copyReleaseData(originalRelease *release.Release, newManifest string, lastDeployed helmv3time.Time) *release.Release {
```

Comment:
I'm a bit confused as to why we needed separate, specific logic for copying this value. Can you explain why we did this as opposed to `newRelease := *originalRelease` (and then modify the fields as necessary on newRelease)?

Reply:
@MbolotSuse this was the context: https://github.com/rancher/rancher/pull/40104#discussion_r1068784419

Reply:
@MbolotSuse I tried changing it to `newRelease := *originalRelease` but it seems it doesn't do a deep copy, so when I change the `newRelease`, I also end up changing `originalRelease` which is undesirable (for the situation when we want to rollback to the previous release, in case of failure).
Could you elaborate on your concerns?